File: android/src/main/java/com/rnmapbox/rnmbx/components/annotation/RNMBXPointAnnotationManager.java
package com.rnmapbox.rnmbx.components.annotation;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.rnmapbox.rnmbx.components.AbstractEventEmitter;
import com.rnmapbox.rnmbx.events.constants.EventKeys;
import com.rnmapbox.rnmbx.utils.GeoJSONUtils;

import java.util.Map;

public class RNMBXPointAnnotationManager extends AbstractEventEmitter<RNMBXPointAnnotation> {
    public static final String REACT_CLASS = "RNMBXPointAnnotation";

    public RNMBXPointAnnotationManager(ReactApplicationContext reactApplicationContext) {
        super(reactApplicationContext);
    }

    @Override
    public Map<String, String> customEvents() {
        return MapBuilder.<String, String>builder()
                .put(EventKeys.POINT_ANNOTATION_SELECTED, "onMapboxPointAnnotationSelected")
                .put(EventKeys.POINT_ANNOTATION_DESELECTED, "onMapboxPointAnnotationDeselected")
                .put(EventKeys.POINT_ANNOTATION_DRAG_START, "onMapboxPointAnnotationDragStart")
                .put(EventKeys.POINT_ANNOTATION_DRAG, "onMapboxPointAnnotationDrag")
                .put(EventKeys.POINT_ANNOTATION_DRAG_END, "onMapboxPointAnnotationDragEnd")
                .build();
    }

    //region React Methods
    public static final int METHOD_REFRESH = 2;

    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        return MapBuilder.<String, Integer>builder()
                .put("refresh", METHOD_REFRESH)
                .build();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected RNMBXPointAnnotation createViewInstance(ThemedReactContext reactContext) {
        return new RNMBXPointAnnotation(reactContext, this);
    }

    @ReactProp(name="id")
    public void setId(RNMBXPointAnnotation annotation, String id) {
        annotation.setID(id);
    }

    @ReactProp(name="coordinate")
    public void setCoordinate(RNMBXPointAnnotation annotation, String geoJSONStr) {
        annotation.setCoordinate(GeoJSONUtils.toPointGeometry(geoJSONStr));
    }

    @ReactProp(name="anchor")
    public void setAnchor(RNMBXPointAnnotation annotation, ReadableMap map) {
        annotation.setAnchor((float) map.getDouble("x"), (float) map.getDouble("y"));
    }

    @ReactProp(name="draggable")
    public void setDraggable(RNMBXPointAnnotation annotation, Boolean draggable) {
        annotation.setDraggable(draggable);
    }

    @Override
    public void receiveCommand(RNMBXPointAnnotation annotation, int commandID, @Nullable ReadableArray args) {
        switch (commandID) {
            case METHOD_REFRESH:
                annotation.refresh();
                break;
        }
    }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/annotation/RNMBXPointAnnotationManager.kt
package com.rnmapbox.rnmbx.components.annotation

import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.RNMBXPointAnnotationManagerDelegate
import com.facebook.react.viewmanagers.RNMBXPointAnnotationManagerInterface
import com.rnmapbox.rnmbx.components.AbstractEventEmitter
import com.rnmapbox.rnmbx.components.styles.sources.RNMBXShapeSource
import com.rnmapbox.rnmbx.events.constants.EventKeys
import com.rnmapbox.rnmbx.events.constants.eventMapOf
import com.rnmapbox.rnmbx.utils.GeoJSONUtils.toPointGeometry
import com.rnmapbox.rnmbx.utils.ViewTagResolver

class RNMBXPointAnnotationManager(reactApplicationContext: ReactApplicationContext, val viewTagResolver: ViewTagResolver) : AbstractEventEmitter<RNMBXPointAnnotation>(reactApplicationContext),
  RNMBXPointAnnotationManagerInterface<RNMBXPointAnnotation> {

    private val mDelegate: ViewManagerDelegate<RNMBXPointAnnotation>

    init {
        mDelegate = RNMBXPointAnnotationManagerDelegate(this)
    }

    override fun getDelegate(): ViewManagerDelegate<RNMBXPointAnnotation> {
        return mDelegate
    }

    override fun getName(): String {
        return REACT_CLASS
    }

    override fun customEvents(): Map<String, String> {
        return eventMapOf(
            EventKeys.POINT_ANNOTATION_SELECTED to "onMapboxPointAnnotationSelected",
            EventKeys.POINT_ANNOTATION_DESELECTED to "onMapboxPointAnnotationDeselected",
            EventKeys.POINT_ANNOTATION_DRAG_START to "onMapboxPointAnnotationDragStart",
            EventKeys.POINT_ANNOTATION_DRAG to "onMapboxPointAnnotationDrag",
            EventKeys.POINT_ANNOTATION_DRAG_END to "onMapboxPointAnnotationDragEnd"
        )
    }

    override fun createViewInstance(reactContext: ThemedReactContext): RNMBXPointAnnotation {
        return RNMBXPointAnnotation(reactContext, this)
    }

    override fun onDropViewInstance(view: RNMBXPointAnnotation) {
        val reactTag = view.id

        viewTagResolver.viewRemoved(reactTag)
        super.onDropViewInstance(view)
    }

    fun tagAssigned(reactTag: Int) {
        return viewTagResolver.tagAssigned(reactTag)
    }

    @ReactProp(name = "id")
    override fun setId(annotation: RNMBXPointAnnotation, id: Dynamic) {
        annotation.iD = id.asString()
    }

    @ReactProp(name = "coordinate")
    override fun setCoordinate(annotation: RNMBXPointAnnotation, geoJSONStr: Dynamic) {
        annotation.setCoordinate(toPointGeometry(geoJSONStr.asString())!!)
    }

    @ReactProp(name = "anchor")
    override fun setAnchor(annotation: RNMBXPointAnnotation, map: Dynamic) {
        annotation.setAnchor(map.asMap().getDouble("x").toFloat(), map.asMap().getDouble("y").toFloat())
    }

    @ReactProp(name = "draggable")
    override fun setDraggable(annotation: RNMBXPointAnnotation, draggable: Dynamic) {
        annotation.setDraggable(draggable.asBoolean())
    }

    companion object {
        const val REACT_CLASS = "RNMBXPointAnnotation"
    }
}